### ISSUE TITLE
Feature Request: open some search links

### ISSUE THREAD

**Issue by iRobie** [sealemar/BingRewards#126]
Would it be possible to randomly open 1-2 links of the search results? The purpose is to make it look like a user is actually clicking links, and therefore isn't a bot. I'd recommend selecting links in a way that advertised links could also be opened, though less frequently.

Thanks!

**Comment by sealemar** [comment]
Have you seen any evidence of Bing! banning an account if it searches, but doesn't do clicks?

**Comment by iRobie** [comment]
I have not. But adding this makes the profile look more real. Your search history does include what link you clicked on, so I know Bing tracks this. I have read online accounts of people being banned for unknown reasons. If I were designing the system (I admit, Microsoft and I don't always see eye to eye), I'd make banning based purely on metrics. The ones I'd start with are active times of day, time between searches, click thru rate, search profile interests.

Again, no evidence this is what bans are based on, I just don't see what bans are actually based on.

You've done a great job at the first two metrics. I've tackled the last one by making a reddit based RSS feed for searches (searches will have roughly consistent themes based on what subreddits you select). 

For the click thru rate, I've written an iMacro that will open your search history, repeat one of your last searches, then open a random search result. I plan to run this side by side with the python script.

The iMacro is very simple: find div with ID b_tween, then click a link after it. However, my Python is very poor (look at my RSS code for evidence of that!). If I do it, I'll check in the code. If you think this is worth it, I suggest you do it instead :)

**Comment by sealemar** [comment]
very good point. Now someone needs to implement that :)

**Comment by iRobie** [comment]
How would you go about this? From my research, this would require mechanism or selenium in order to 'click' a link, rather than just opening the URL. So, rewriting a lot of the current code to switch the browser portion.

Can you think of a better method before I start? There's a selenium bing project up on Github so it wouldn't be starting from nothing, but again I warn you my python is very weak!

**Comment by amayer5125** [comment]
I am a web developer and avid python user. I believe just going to the link is sufficient enough. I dont believe you actually need to click the link.
The server (bing database) is only aware of what pages you visit not what you physically clicked on to get there.
So this should be fairly simple to implement. I can look into it later this week or early next week.

**Comment by amayer5125** [comment]
One way we will be able to tell if it is working is by checking the search history clicks like you mentioned.

**Comment by iRobie** [comment]
Wow, you are right. I tested the following way:
Performed a search on computer A
Copied search result link from computer A, pasted into Notepad, then pasted into browser (no tracking tags)

On computer B, check search history (same account)
Verified Bing can see that I opened that link after searching on a term


Well, that simplifies this!

**Comment by amayer5125** [comment]
@iRobie It should be fairly simple. Ive thought about this before but since nobody was getting banned for it (to my knowledge) I didnt see an big need to implement it right way.

**Comment by sealemar** [comment]
@amayer5125 I'm with you on that. It's good to be proactive and have that implemented before Bing starts banning for that (if ever), but no evidence of a ban has been noticed so far thus a change even though a simple one will require some time which I would be glad if someone could invest.

**Comment by iRobie** [comment]
Without the need for selenium, this is more reasonable than I thought. I'm trying to parse the results with beautifulsoup4, which I've got working in a sandbox already. Need to integrate it into the rest of the project.

**Comment by amayer5125** [comment]
@iRobie we just regex or search the code for the links. We dont need to use beautifulsoup or anything. We arent parsing and checking structures we just need to get the links from the page.

**Comment by iRobie** [comment]
I got this working. I'll check in my code tomorrow, after some cleanup on it! I did use BeautifulSoup, but I'm sure it could be done in regex too.

Basically, Bing isn't checking for a click. Instead, on some event on the link (right click works, which is why my previous test worked), it fires a call to GLinkPing.aspx with your session ID (I believe) and the link you're interacting with. Sending that call adds the link to your history, which is what I got working.

I also saw more calls - repeated calls to lsp.aspx which I believe tracks the timing on a webpage, and something else. You can see a sample here: http://webmasters.stackexchange.com/questions/65577/how-does-bing-know-when-a-result-is-clicked-to-use-in-website-ranking

The interesting data is in ClientInstRequest. IG is your session ID, TS and STS are timing, then the D is the something else

Automating that is out of the scope of this request!

**Comment by sealemar** [comment]
Please, don't add extra dependencies which [have to be installed separately](http://www.crummy.com/software/BeautifulSoup/#Download).

```pkg.bingRewards.BingRewards.opener``` is all you need. It keeps the cookies. Headers are added separately. See [this](https://github.com/sealemar/BingRewards/blob/master/pkg/bingRewards.py#L101) for example usage:

```python
def requestFlyoutPage(self):
    """
    Returns bing.com flyout page
    This page shows what rewarding activity can be performed in
    order to earn Bing points
    """
    url = self.BING_FLYOUT_PAGE
    request = urllib2.Request(url = url, headers = self.httpHeaders)

    with self.opener.open(request) as response:
        page = helpers.getResponseBody(response)
    return page
```

With that ```opener``` you need to open a url _after_ you completed a corresponding search with that opener, thus all your cookies and stuff is preserved.

To have a more complete understanding of how it works I suggest you to read how stateless HTTP protocol can preserve some notion of state through cookies.

**Comment by amayer5125** [comment]
@sealemar I think @iRobie is using BeautifulSoup to parse the page not download it. This is why I recommended using regex to get the links.

**Comment by iRobie** [comment]
Feel free to close this request. I got it working in this commit, which admittingly is an inefficient way to do it: https://github.com/iRobie/BingRewards/commit/9a638d4e4f9cf8d1fbea948e79b438fcd1355fc1

The clicked links are not tracked by cookies or headers. It's tracked by an AJAX call. BeautifulSoup certainly isn't required - all that you need is the IG string and search results ID. I'm sure regex can pull that info.

More code is needed than I did, like error handling.

**Comment by amayer5125** [comment]
@iRobie @sealemar This issue can be closed. This feature was implemented in facd7c0303b74745c82a6c0693a522b1e21d736c
